clang-tools-extra/clang-doc/Generators.h

```diff
@@ -52,6 +54,85 @@ findGeneratorByName(llvm::StringRef Format);
+        llvm::MemoryBuffer::getFile(FileName);
+    if (auto EC = BufferOrError.getError())
+      return createFileOpenError(FileName, EC);
+    return std::make_unique<MustacheTemplateFile>(
+        std::move(BufferOrError.get()));
+  }
+
+  llvm::Error registerPartialFile(StringRef Name, StringRef FileName) {
+    llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> BufferOrError =
+        llvm::MemoryBuffer::getFile(FileName);
+    if (auto EC = BufferOrError.getError())
+      return createFileOpenError(FileName, EC);
+
+    std::unique_ptr<llvm::MemoryBuffer> Buffer = std::move(BufferOrError.get());
+    StringRef FileContent = Buffer->getBuffer();
+    T.registerPartial(Name.str(), FileContent.str());
+    return llvm::Error::success();
+  }
+
+  void render(llvm::json::Value &V, raw_ostream &OS) { T.render(V, OS); }
+
+  MustacheTemplateFile(std::unique_ptr<llvm::MemoryBuffer> &&B)
+      : Saver(Allocator), Ctx(Allocator, Saver), T(B->getBuffer(), Ctx),
+        Buffer(std::move(B)) {}
+};
+
+struct MustacheGenerator {
+  Expected<std::string> getInfoTypeStr(llvm::json::Object *Info,
+                                       StringRef Filename);
+
+  /// Used to find the relative path from the file to the format's docs root.
+  /// Mainly used for the HTML resource paths.
+  SmallString<128> getRelativePathToRoot(StringRef PathToFile,
+                                         StringRef DocsRootPath);
+  virtual ~MustacheGenerator() = default;
+
+  /// Initializes the template files from disk and calls setupTemplate to
+  /// register partials
+  virtual llvm::Error setupTemplateFiles(const ClangDocContext &CDCtx) = 0;
+
+  /// Populates templates with data from JSON and calls any specifics for the
+  /// format. For example, for HTML it will render the paths for CSS and JS.
+  virtual llvm::Error generateDocForJSON(llvm::json::Value &JSON,
+                                         llvm::raw_fd_ostream &OS,
+                                         const ClangDocContext &CDCtx,
+                                         StringRef ObjectTypeStr,
+                                         StringRef RelativeRootPath) = 0;
+
+  /// Registers partials to templates.
+  llvm::Error
+  setupTemplate(std::unique_ptr<MustacheTemplateFile> &Template,
+                StringRef TemplatePath,
+                std::vector<std::pair<StringRef, StringRef>> Partials);
+
+  /// \brief The main orchestrator for Mustache-based documentation.
+  ///
+  /// 1. Initializes templates files from disk by calling setupTemplateFiles.
+  /// 2. Calls the JSON generator to write JSON to disk.
+  /// 3. Iterates over the JSON files, recreates the directory structure from
+  /// JSON, and calls generateDocForJSON for each file.
+  /// 4. A file of the desired format is created.
+  llvm::Error generateDocumentation(
+      StringRef RootDir, llvm::StringMap<std::unique_ptr<doc::Info>> Infos,
+      const clang::doc::ClangDocContext &CDCtx, std::string DirName);
```

Comment:
Having both `generateDocumentation` and `generateDocs` as methods in the same class is a bit confusing. Would it make sense for `generateDocumentation` to be `virtual` and `generateDocs` and `MustacheHTMLGenerator` to be the overridden version?

Reply:
you could update the `GenerateDocs()` signature to have a trailing string argument w/ a default value, right? then there isn't a signature mismatch.